include/ppx/geometry.h

```diff
@@ -165,35 +183,59 @@ class Geometry
         {
             uint32_t sizeOfValue = static_cast<uint32_t>(sizeof(T));
 
-            // Current size
-            size_t offset = mData.size();
-            // Allocate storage for incoming data
-            mData.resize(offset + sizeOfValue);
+            if (mKnownElementCount == 0) {
+                // Allocate storage for incoming data
+                SetSize(mOffset + sizeOfValue);
+            }
+
             // Copy data
             const void* pSrc = &value;
-            void*       pDst = mData.data() + offset;
+            void*       pDst = mData.data() + mOffset;
             memcpy(pDst, pSrc, sizeOfValue);
+
+            mOffset += sizeOfValue;
         }
 
         template <typename T>
         void Append(uint32_t count, const T* pValues)
         {
             uint32_t sizeOfValues = count * static_cast<uint32_t>(sizeof(T));
 
-            // Current size
-            size_t offset = mData.size();
-            // Allocate storage for incoming data
-            mData.resize(offset + sizeOfValues);
+            if (mKnownElementCount == 0) {
+                // Allocate storage for incoming data
+                SetSize(mOffset + sizeOfValues);
+            }
+
             // Copy data
             const void* pSrc = pValues;
-            void*       pDst = mData.data() + offset;
+            void*       pDst = mData.data() + mOffset;
             memcpy(pDst, pSrc, sizeOfValues);
+
+            mOffset += sizeOfValues;
+        }
+
+        // If Overwrite is used to write new data, calling code will also
+        // have to SetOffset() appropriately
+        //
+        template <typename T>
+        void Overwrite(const T& value, size_t offset)
+        {
+            uint32_t sizeOfValue = static_cast<uint32_t>(sizeof(T));
+
+            PPX_ASSERT_MSG(offset + sizeOfValue <= GetSize(), "Attempting to overwrite with value of size " << sizeOfValue << " at offset " << offset << " will overflow the buffer of size " << GetSize());
+
+            // Copy data
+            const void* pSrc = &value;
+            void*       pDst = mData.data() + offset;
+            memcpy(pDst, pSrc, sizeOfValue);
         }
 
     private:
-        BufferType        mType        = BUFFER_TYPE_VERTEX;
-        uint32_t          mElementSize = 0;
+        BufferType        mType              = BUFFER_TYPE_VERTEX;
+        uint32_t          mElementSize       = 0;
+        uint32_t          mKnownElementCount = 0;
```

Comment:
Can this be `mMaxElementCount` to keep the names across the various classes?

Reply:
Done